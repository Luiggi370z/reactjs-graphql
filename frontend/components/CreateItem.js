import React, { Component } from 'react'
import { Mutation } from 'react-apollo'
import Router from 'next/router'
import gql from 'graphql-tag'
import Form from './styles/Form'
import Error from './ErrorMessage'
import formatMoney from '../lib/formatMoney'

const CREATE_ITEM_MUTATION = gql`
	mutation CREATE_ITEM_MUTATION(
		$title: String!
		$description: String!
		$price: Int!
		$image: String
		$largeImage: String
	) {
		createItem(
			title: $title
			description: $description
			price: $price
			image: $image
			largeImage: $largeImage
		) {
			id
		}
	}
`

class CreateItem extends Component {
	state = {
		title: 'aaaa',
		description: 'bbbbbb',
		image: '',
		largeImage: '',
		price: 0
	}

	render() {
		return (
			<Mutation mutation={CREATE_ITEM_MUTATION} variables={this.state}>
				{(createItem, { loading, error }) => (
					<Form
						onSubmit={async e => {
							e.preventDefault()
							const res = await createItem()
							console.log(res)
							Router.push({
								pathname: '/item',
								query: { id: res.data.createItem.id }
							})
						}}>
						<Error error={error} />
						<fieldset disabled={loading} aria-busy={loading}>
							<label htmlFor='image'>
								Image
								<input
									id='image'
									type='file'
									name='image'
									id='image'
									placeholder='Select an image'
									required
									onChange={this.uploadFile}
								/>
								{this.state.image && (
									<img src={this.state.image} alt='Upload preview' />
								)}
							</label>
							<label htmlFor='title'>
								Title
								<input
									id='title'
									type='text'
									name='title'
									id='title'
									placeholder='Title'
									required
									value={this.state.title}
									onChange={this.handleChange}
								/>
							</label>
							<label htmlFor='price'>
								Price
								<input
									id='price'
									type='number'
									name='price'
									id='price'
									placeholder='Price'
									required
									value={this.state.price}
									onChange={this.handleChange}
								/>
							</label>
							<label htmlFor='description'>
								description
								<textarea
									name='description'
									id='description'
									placeholder='Enter a description'
									required
									value={this.state.description}
									onChange={this.handleChange}
								/>
							</label>
							<button type='submit'>Submit</button>
						</fieldset>
					</Form>
				)}
			</Mutation>
		)
	}

	handleChange = e => {
		const { name, type, value } = e.target
		const val = type === 'number' ? parseFloat(value) : value
		this.setState({ [name]: val })
	}

	uploadFile = async e => {
		console.log('Uploading file...')

		const files = e.target.files
		const data = new FormData()
		data.append('file', files[0])
		data.append('upload_preset', 'sickfits')

		const res = await fetch(
			'https://api.cloudinary.com/v1_1/luiggiamg/image/upload',
			{
				method: 'POST',
				body: data
			}
		)

		const file = await res.json()
		console.log(file)
		this.setState({
			image: file.secure_url,
			largeImage: file.eager[0].secure_url
		})
	}
}

export default CreateItem
export { CREATE_ITEM_MUTATION }
