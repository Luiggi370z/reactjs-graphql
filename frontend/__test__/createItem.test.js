import wait from 'waait'
import { mount } from 'enzyme'
import toJSON from 'enzyme-to-json'
import Router from 'next/router'
import { MockedProvider } from 'react-apollo/test-utils'
import CreateItem, { CREATE_ITEM_MUTATION } from '../components/CreateItem'
import { fakeItem } from '../lib/testUtils'

const dogImage = 'https://dog.com/dog.jpg'

// mock the global fetch API
global.fetch = jest.fn().mockResolvedValue({
	json: () => ({
		secure_url: dogImage,
		eager: [{ secure_url: dogImage }]
	})
})

describe('<CreateItem/>', () => {
	test('should render and matches snapshot', async () => {
		const wrapper = mount(
			<MockedProvider>
				<CreateItem />
			</MockedProvider>
		)

		const form = wrapper.find('form[data-test="form"]') //tag form added to avoid double Form in the snapshot Form component and form tag
		expect(toJSON(form)).toMatchSnapshot()
	})

	test('should uploads a file when changed', async () => {
		const wrapper = mount(
			<MockedProvider>
				<CreateItem />
			</MockedProvider>
		)

		const input = wrapper.find('input[type="file"]') //tag form added to avoid double Form in the snapshot Form component and form tag
		input.simulate('change', { target: { files: ['fakedog.jpg'] } })

		await wait()

		const component = wrapper.find('CreateItem').instance()
		expect(component.state.image).toEqual(dogImage)
		expect(component.state.largeImage).toEqual(dogImage)
		expect(global.fetch).toHaveBeenCalled()

		global.fetch.mockReset()
	})

	test('should udpate state', async () => {
		const wrapper = mount(
			<MockedProvider>
				<CreateItem />
			</MockedProvider>
		)

		wrapper
			.find('#title')
			.simulate('change', { target: { value: 'Testing', name: 'title' } })
		wrapper.find('#price').simulate('change', {
			target: { value: 50000, name: 'price', type: 'number' }
		})
		wrapper.find('#description').simulate('change', {
			target: { value: 'Test description', name: 'description' }
		})

		expect(wrapper.find('CreateItem').instance().state).toMatchObject({
			title: 'Testing',
			price: 50000,
			description: 'Test description'
		})
	})

	test('should udpate state', async () => {
		const item = fakeItem()
		const mocks = [
			{
				request: {
					query: CREATE_ITEM_MUTATION,
					variables: {
						title: item.title,
						description: item.description,
						image: '',
						largeImage: '',
						price: item.price
					}
				},
				result: {
					data: {
						createItem: {
							...fakeItem,
							id: '123', // required field!
							__typename: 'Item'
						}
					}
				}
			}
		]

		const wrapper = mount(
			<MockedProvider mocks={mocks}>
				<CreateItem />
			</MockedProvider>
		)

		wrapper
			.find('#title')
			.simulate('change', { target: { value: item.title, name: 'title' } })
		wrapper.find('#price').simulate('change', {
			target: { value: item.price, name: 'price', type: 'number' }
		})
		wrapper.find('#description').simulate('change', {
			target: { value: item.description, name: 'description' }
		})

		//mock router
		Router.router = { push: jest.fn() }

		wrapper.find('form').simulate('submit')

		await wait(50)

		expect(Router.router.push).toHaveBeenCalled()
		expect(Router.router.push).toHaveBeenCalledWith({
			pathname: '/item',
			query: { id: '123' }
		})
	})
})
