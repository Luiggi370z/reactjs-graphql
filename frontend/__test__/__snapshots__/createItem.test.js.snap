// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateItem/> should render and matches snapshot 1`] = `
<form
  className="Form-sc-1kmq3n-0 gEujYN"
  data-test="form"
  onSubmit={[Function]}
>
  <DisplayError
    error={Object {}}
  />
  <fieldset
    aria-busy={false}
    disabled={false}
  >
    <label
      htmlFor="image"
    >
      Image
      <input
        id="image"
        name="image"
        onChange={[Function]}
        placeholder="Select an image"
        required={true}
        type="file"
      />
    </label>
    <label
      htmlFor="title"
    >
      Title
      <input
        id="title"
        name="title"
        onChange={[Function]}
        placeholder="Title"
        required={true}
        type="text"
        value="aaaa"
      />
    </label>
    <label
      htmlFor="price"
    >
      Price
      <input
        id="price"
        name="price"
        onChange={[Function]}
        placeholder="Price"
        required={true}
        type="number"
        value={0}
      />
    </label>
    <label
      htmlFor="description"
    >
      description
      <textarea
        id="description"
        name="description"
        onChange={[Function]}
        placeholder="Enter a description"
        required={true}
        value="bbbbbb"
      />
    </label>
    <button
      type="submit"
    >
      Submit
    </button>
  </fieldset>
</form>
`;
